stanford_soe_r25.pages.inc

```diff
@@ -0,0 +1,994 @@
+        $attr_temp = str_replace('[r25_attr_id]', $key, $attr_str);
+        $attr_temp = str_replace('[r25_attr_type]', $value['type'], $attr_temp);
+        $attr_temp = str_replace('[r25_attr_value]', $form_state['values']['stanford_soe_r25_booking_attr' . $key], $attr_temp);
+        $attr_insert .= $attr_temp;
+      }
+    }
+  }
+
+  // if there is a 25Live custom attribute we've defined for contact information, add the XML snippet for it to the request
+  if (!empty($room['contact_attr_field'])) {
+    foreach ($room['contact_attr_field'] as $key => $value) {
+      if (!empty($form_state['values']['stanford_soe_r25_contact_' . $key])) {
+        $attr_temp = str_replace('[r25_attr_id]', $key, $attr_str);
+        $attr_temp = str_replace('[r25_attr_type]', $value['type'], $attr_temp);
+        $attr_temp = str_replace('[r25_attr_value]', $form_state['values']['stanford_soe_r25_contact_' . $key], $attr_temp);
+        $attr_insert .= $attr_temp;
+      }
+    }
+  }
+
+  // get the XML template for creating an event and replace tokens with data for this reservation
+  $event_state = $event_state - 1;
+  $xml_file = '/stanford_soe_r25_reserve.xml';
+  $xml = file_get_contents(drupal_get_path('module', 'stanford_soe_r25') . $xml_file); //'/stanford_soe_r25_reserve.xml');
+  $xml = str_replace('[r25_event_name]', $form_state['values']['stanford_soe_r25_booking_reason'], $xml);
+  $xml = str_replace('[r25_parent_id]', variable_get('stanford_soe_r25_parent_event_id', 'unknown'), $xml);
+  $xml = str_replace('[r25_event_type]', variable_get('stanford_soe_r25_event_type', 'unknown'), $xml);
+  $xml = str_replace('[r25_event_state]', $event_state, $xml);
+  $xml = str_replace('[r25_organization_id]', variable_get('stanford_soe_r25_org_id', 'unknown'), $xml);
+  $xml = str_replace('[r25_expected_headcount]', $form_state['complete form']['stanford_soe_r25_booking_headcount']['#options'][$form_state['values']['stanford_soe_r25_booking_headcount']], $xml);
+  $xml = str_replace('[r25_start_date_time]', $form_state['storage']['stanford_soe_r25_booking']['dates']['start'], $xml);
+  $xml = str_replace('[r25_end_date_time]', $form_state['storage']['stanford_soe_r25_booking']['dates']['end'], $xml);
+  $xml = str_replace('[r25_space_id]', $form_state['storage']['stanford_soe_r25_booking']['room']['space_id'], $xml);
+  $xml = str_replace('[r25_todo]', $todo_insert, $xml);
+  $xml = str_replace('[r25_attr]', $attr_insert, $xml);
+
+  // we want to put some information about the user making this request into the event description to be displayed on the calendar
+  global $user;
+  $res_username = '';
+  $res_usermail = '';
+  if ($user->uid == 0) {
+    if (!empty($form_state['input']['external_username'])) {
+      $res_username = $form_state['input']['external_username'];
+    }
+    if (!empty($form_state['input']['external_usermail'])) {
+      $res_usermail = $form_state['input']['external_usermail'];
```

Comment:
$form_state['input'] should not be used in submit functions as it contains raw unvalidated submissions.
https://drupal.stackexchange.com/questions/112312/difference-between-form-statevalues-and-form-stateinput